backend/src/gdc/gdc.ts

```diff
@@ -0,0 +1,320 @@
+
+import {
+  Goal,
+  Dataseries,
+  GDCOutput,
+  IndicatorWithoutGoal,
+  YearlyGrowth,
+  CumulativeScore,
+  Score,
+  IndicatorScore,
+} from '../types/gdcTypes';
+
+export const computeGDC = (
+  dataseries: Dataseries[],
+  goalArray: Goal[],
+  historicalData: Dataseries[],
+): GDCOutput => {
+  const goals: Map<string, Goal> = new Map<string, Goal>();
+
+  goalArray.forEach((goal) => {
+    const isVariant = goal.dataseries !== undefined;
+    const displayKPI = goal.kpi + (isVariant ? ` - ${goal.dataseries}` : '');
+    goals.set(displayKPI, goal);
+  });
+
+  const outputIndicatorScores = new Map<string, IndicatorScore>();
+  const outputCategoryScores = new Map<string, Score>();
+  const outputSubdomainScores = new Map<string, Score>();
+  const outputDomainScores = new Map<string, Score>();
+
+  const indicatorsWithoutGoals = new Map<string, IndicatorWithoutGoal>();
+  const unreportedIndicators = new Set(u4sscKpis);
+
+  const categoryScores = new Map<string, IndicatorScore[]>();
+  const subdomainScores = new Map<string, CumulativeScore[]>();
+  const domainScores = new Map<string, CumulativeScore[]>();
+
+  dataseries.forEach((series) => {
+    const isVariant = series.dataseries !== undefined;
+    const displayKPI = series.kpi + (isVariant ? ` - ${series.dataseries}` : '');
+
+    const goal = goals.get(displayKPI);
+    unreportedIndicators.delete(series.kpi);
+
+    if (goal === undefined) {
+      indicatorsWithoutGoals.set(displayKPI, {
+        kpi: series.kpi,
+        dataseries: series.dataseries,
+        historicalData: [],
+        yearlyGrowth: [],
+        currentCAGR: 0.0,
+        trendMean: 0.0,
+        trendStd: 0.0,
+        calculationMethod: series.calculationMethod,
+      });
+
+      return;
+    }
+
+    const score = computeScore(series.kpi, series, goal);
+
+    outputIndicatorScores.set(displayKPI, score);
+
+    const category = u4sscKpiToCategory.get(series.kpi);
+    if (category === undefined) throw new ApiError(400, 'WUT');
+
+    const arr = categoryScores.get(category);
+    if (arr === undefined) categoryScores.set(category, [score]);
+    else if (series.dataseries !== undefined) arr.push(score);
+  });
+
+  // aggregate historical data
+
+  historicalData.forEach((hist) => {
+    const isVariant = hist.dataseries !== undefined;
+    const displayKPI = hist.kpi + (isVariant ? ` - ${hist.dataseries}` : '');
+
+    let score: IndicatorScore | IndicatorWithoutGoal | undefined = outputIndicatorScores.get(
+      displayKPI,
+    );
+    if (score === undefined) {
+      // Did not get score from indicators with goals, try the ones without goals instead...
+      score = indicatorsWithoutGoals.get(displayKPI);
+
+      if (score === undefined) return;
+    }
+
+    score.historicalData.push({ year: hist.year, value: hist.value });
+  });
+
+  // calculate statistical data
+  // Need to use a for-of loop as we're assigning to the properties of the iterator variable
```

Comment:
I see this comment in mulitple files, but I'm not sure I understand why. The subject in forEach() is still mutable.